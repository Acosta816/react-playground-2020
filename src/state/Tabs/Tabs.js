import React from 'react';
import './Tabs.css';


class Tabs extends React.Component {

    static defaultProps = {
        tabs: []
    }

    state = {
        currentIndex: 0
    }

    //On an event handler, if you plan on passing a value through to a method, you must place it into an "anonumous arrow function" like so... ()=> {this.handleClick(juice)}
    //and of course all of this inside the onClick={}  ...so it'll look like this... onClick={ ()=> {this.handleClick(apples)} }
    renderButtons = ()=> {
        const tabsButtons = this.props.tabs.map((tab,index)=> {
            return <button onClick={()=> {this.handleClick(index)}} key={index}>{tab.name}</button>
        });

        return tabsButtons;
    }

    renderContent = ()=> {
        const currentTab = this.props.tabs[this.state.currentIndex];
        
        return (
            <div className='content'>
                {currentTab.content}
                <img className="avatar" src={currentTab.avatar} alt="sonic characters" />
            </div>
        )
    }

    handleClick = (index)=> {
        console.log(index);

        this.setState({
            currentIndex: index
        })
    }

    render() {
        console.log(this.props.tabs);

        //Below you may have noticed that the method calls have the parens at the end like so... "this.renderButtons()"  
        //But when we called class arrow methods in eventHandlers we didnt need to add the (parens). Well this is because with event handlers, 
        //we are NOT calling it on the spot inside the handler when we type it in. It's just a reference to it, a painting of it. The 'onClick' code then
        //finds the actual method, wherever it is that you declard it, and then activates it. All we are doing is inserting a variable tag basically. 
        //We are just passing the variable name of the method like so.. onClick={this.renderButtons}   exactly like an argument. The (parens) signify a call on the spot.
        return(
            <div>
                {this.renderButtons()}
                {this.props.tabs.length ? this.renderContent() : null}
                
            </div>
        )
    }
}

export default Tabs;