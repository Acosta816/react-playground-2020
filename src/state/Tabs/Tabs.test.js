import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import Tabs from './Tabs';


describe('ðŸ¦¾TestSuite: Tabs Component', ()=> {
    //variables for below tests
    const tabsProp = [{name:'Sonic', content: 'Fastest thing Alive'}, {name:'Tails', content: 'Loyal and Hyper-intelligent Flying Fox'}, {name:'Amy', content: 'Sonics Biggest Fan. Hammer-weilding Pink Hedgehog'}, {name:'Robotnik', content: 'Crazed Genius, obsessed with robotisizing all life and ridding the world of Sonic'}, {name:'Shadow', content: 'Sonics adversary. Created by Robotnik to become the ultimate lifeform'}, {name:'Sally', content: 'Rightful Princess of Mobius and one of the Leaders in the Freedom Fighters against Robotnik. Sonics Best friend.'}];


    it('ðŸ”¥SMOKE: renders without errors', ()=> {
        const div = document.createElement('div');

        ReactDOM.render(<Tabs/>, div);

        ReactDOM.unmountComponentAtNode(div);

    });

    it('ðŸ“·EnzSnap#01: renders empty given no tabs props passed in', () => {
        const wrapper = shallow(<Tabs />);
        expect(toJSON(wrapper)).toMatchSnapshot();
    });

    it('ðŸ“·EnzSnap#02: renders first tab by default', ()=> {
        const wrapper = shallow(<Tabs tabs={tabsProp} />)
        expect(toJSON(wrapper)).toMatchSnapshot();
    });

    //Enzyme exports a method called "shallow" which allows you to wrap your isolated component is a 'jquery-like' wrapper to render the component by itself.
    //This is known as shallow-rendering (rendering just the component without the children). This was difficult to do before enzyme. once the wrapper is created, we reference that 
    //wrapper to do different thing with it like traverse the component to find perhaps a certain button using css selector just like jquery, like so... wrapper.find('button').at(1).simulate('click');
    //We found the 'button' at index#1 and simulated a 'click' on that button.

    it('ðŸ¦ ðŸ’¥EnzEvent#01: closes first tab and opens any clicked tab', ()=> {
        const wrapper= shallow(<Tabs tabs={tabsProp} />); 
        wrapper.find('button').at(1).simulate('click');
        expect(toJSON(wrapper)).toMatchSnapshot();

        console.log('>>> WRAPPER <<<');
        console.log(wrapper.debug());
        console.log('>>> FIND(BUTTON) <<<');
        console.log(wrapper.find('button').at(1).debug());
    })



})