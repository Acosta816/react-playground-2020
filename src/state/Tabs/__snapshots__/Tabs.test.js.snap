// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ðŸ¦¾TestSuite: Tabs Component ðŸ“·EnzSnap#01: renders empty given no tabs props passed in 1`] = `<div />`;

exports[`ðŸ¦¾TestSuite: Tabs Component ðŸ“·EnzSnap#02: renders first tab by default 1`] = `
<div>
  <button
    key="0"
    onClick={[Function]}
  >
    Sonic
  </button>
  <button
    key="1"
    onClick={[Function]}
  >
    Tails
  </button>
  <button
    key="2"
    onClick={[Function]}
  >
    Amy
  </button>
  <button
    key="3"
    onClick={[Function]}
  >
    Robotnik
  </button>
  <button
    key="4"
    onClick={[Function]}
  >
    Shadow
  </button>
  <button
    key="5"
    onClick={[Function]}
  >
    Sally
  </button>
  <div
    className="content"
  >
    Fastest thing Alive
    <img
      alt="sonic characters"
      className="avatar"
    />
  </div>
</div>
`;

exports[`ðŸ¦¾TestSuite: Tabs Component ðŸ¦ ðŸ’¥EnzEvent#01: closes first tab and opens any clicked tab 1`] = `
<div>
  <button
    key="0"
    onClick={[Function]}
  >
    Sonic
  </button>
  <button
    key="1"
    onClick={[Function]}
  >
    Tails
  </button>
  <button
    key="2"
    onClick={[Function]}
  >
    Amy
  </button>
  <button
    key="3"
    onClick={[Function]}
  >
    Robotnik
  </button>
  <button
    key="4"
    onClick={[Function]}
  >
    Shadow
  </button>
  <button
    key="5"
    onClick={[Function]}
  >
    Sally
  </button>
  <div
    className="content"
  >
    Loyal and Hyper-intelligent Flying Fox
    <img
      alt="sonic characters"
      className="avatar"
    />
  </div>
</div>
`;
